File: array/pr2/rotate_array.cpp
#include <string>
#include <vector>

using namespace std;

int dx[4] = {0, 1, 0, -1};
int dy[4] = {1, 0, -1, 0};

int rotate_array(vector<int> query, vector<vector<int>> &matrix)
{
  int x = query[0] - 1;
  int y = query[1] - 1;

  int curr_num = matrix[x][y];
  int minv = curr_num;

  int dir = 0;
  while (dir < 4)
  {
    int nx = x + dx[dir];
    int ny = y + dy[dir];
    if (nx < query[0] - 1 || nx > query[2] - 1 || ny < query[1] - 1 || ny > query[3] - 1)
    {
      dir += 1;
      continue;
    }
    int temp = matrix[nx][ny];
    matrix[nx][ny] = curr_num;
    curr_num = temp;
    x = nx;
    y = ny;
    minv = min(minv, temp);
  }
  return minv;
}
vector<int> solution(int rows, int columns, vector<vector<int>> queries)
{
  vector<int> answer;
  vector<vector<int>> matrix(rows, vector<int>(columns));
  for (int i = 1; i <= rows; i++)
  {
    for (int j = 1; j <= columns; j++)
    {
      matrix[i - 1][j - 1] = j + (i - 1) * columns;
    }
  }
  for (auto query : queries)
  {
    int minv = rotate_array(query, matrix);
    answer.push_back(minv);
  }
  return answer;
}

File: array/pr2/rotate_array.js
function rotateArray(query, matrix) {
  //set clockwise 약
  let dx = [0, 1, 0, -1];
  let dy = [1, 0, -1, 0];

  let x = query[0] - 1;
  let y = query[1] - 1;

  let currNum = matrix[x][y];
  let minVal = currNum;

  let dir = 0;
  while (dir < 4) {
    let nx = x + dx[dir];
    let ny = y + dy[dir];

    if (nx < query[0] - 1 || nx > query[2] - 1 || ny < query[1] - 1 || ny > query[3] - 1) {
      dir++;
      continue;
    }

    let temp = matrix[nx][ny];
    matrix[nx][ny] = currNum;
    currNum = temp;
    x = nx;
    y = ny;
    minVal = Math.min(minVal, temp);
  }
  return minVal;
}

function solution(rows, columns, queries) {
  let answer = [];
  let matrix = Array.from({ length: rows }, (_, i) =>
    Array.from({ length: columns }, (_, j) => j + 1 + i * columns)
  );

  queries.forEach(query => {
    let minValue = rotateArray(query, matrix);
    answer.push(minValue);
  });

  return answer;
}
